x/ccv/provider/keeper/key_assignment.go

```diff
@@ -0,0 +1,739 @@
+	// Delete the current mapping from the consumer key to the provider key
+	if ck, ok := ka.Store.GetProviderConsAddrToConsumerPublicKey(pca); ok {
+		ka.Store.DelConsumerPublicKeyToProviderPublicKey(ck)
+	}
+	// Delete the current mapping from the provider key to the consumer key
+	ka.Store.DelProviderConsAddrToConsumerPublicKey(pca)
+	toDelete := []ConsumerConsAddr{}
+	// Find all the consumer keys which were mapped to by the provider key
+	// in order to delete them
+	ka.Store.IterateConsumerConsAddrToLastUpdateMemo(func(cca ConsumerConsAddr, lum providertypes.LastUpdateMemo) bool {
+		pcaInMemo := TMCryptoPublicKeyToConsAddr(*lum.ProviderKey)
+		if pca.Equals(pcaInMemo) {
+			toDelete = append(toDelete, cca)
+		}
+		return false
+	})
+	// Delete the mappings from the consumer keys to the provider keys
+	for _, cca := range toDelete {
+		ka.Store.DelConsumerConsAddrToLastUpdateMemo(cca)
+	}
+	return nil
+}
+
+// GetCurrentConsumerPubKeyFromProviderPubKey returns the current consumer key assigned to the provider key.
+func (ka *KeyAssignment) GetCurrentConsumerPubKeyFromProviderPubKey(pk ProviderPublicKey) (ck ConsumerPublicKey, found bool) {
+	return ka.Store.GetProviderConsAddrToConsumerPublicKey(TMCryptoPublicKeyToConsAddr(pk))
+}
+
+// GetProviderPubKeyFromConsumerPubKey returns any provider key which is currently assigned to the consumer key.
+func (ka *KeyAssignment) GetProviderPubKeyFromConsumerPubKey(ck ConsumerPublicKey) (pk ProviderPublicKey, found bool) {
+	return ka.Store.GetConsumerPublicKeyToProviderPublicKey(ck)
+}
+
+// GetProviderPubKeyFromConsumerConsAddress returns the provider key which was associated to the consumer key with
+// consensus address cca at the last update that used the consumer key.
+func (ka *KeyAssignment) GetProviderPubKeyFromConsumerConsAddress(cca sdk.ConsAddress) (pk ProviderPublicKey, found bool) {
+	if lum, found := ka.Store.GetConsumerConsAddrToLastUpdateMemo(cca); found {
+		return *lum.ProviderKey, true
+	}
+	return pk, false
+}
+
+// PruneUnusedKeys deletes from the store all of the consumer keys which the consumer chain has matured, based
+// on latestMatureVscid, and for which it sure that the correct consumer chain can no longer reference the key
+// in a downtime or double sign slash instruction.
+func (ka *KeyAssignment) PruneUnusedKeys(latestMatureVscid VSCID) {
+	toDel := []ConsumerConsAddr{}
+	ka.Store.IterateConsumerConsAddrToLastUpdateMemo(func(cca ConsumerConsAddr, lum providertypes.LastUpdateMemo) bool {
+		if lum.Power == 0 && lum.Vscid <= latestMatureVscid {
+			toDel = append(toDel, cca)
+		}
+		return false
+	})
+	for _, cca := range toDel {
+		ka.Store.DelConsumerConsAddrToLastUpdateMemo(cca)
+	}
+}
+
+// DeterministicStringify returns a deterministic string representation of the
+// key. This is useful to identify like keys with different representations.
+func DeterministicStringify(k tmprotocrypto.PublicKey) string {
```

Comment:
Marshall serializes to bytes such that ANY protobuf compatible language can deserialize the bytes to a native representation if they are aware of the protobuf definition. So I don't think Marshalled data contains anything to do with Go.

You can't use tmprotocrypto.PublicKey as a golang map key because it is a struct with an interface inside

https://github.com/cosmos/interchain-security/blob/4444034d1dfa72328fd7ac2a8bf38915933b5a67/x/ccv/provider/keeper/key_assignment_test.go#L521